lib/spack/spack/solver/concretize.lp

```diff
@@ -36,16 +36,54 @@ version_satisfies(Package, Constraint)
 #defined preferred_version_declared/3.
 #defined version_satisfies/3.
 
+%-----------------------------------------------------------------------------
+% Spec conditions and imposed constraints
+%
+% Given Spack directives like these:
+%    depends_on("foo@1.0+bar", when="@2.0+variant")
+%    provides("mpi@2:", when="@1.9:")
+%
+% The conditions are `@2.0+variant` and `@1.9:`, and the imposed constraints
+% are `@1.0+bar` on `foo` and `@2:` on `mpi`.
+%-----------------------------------------------------------------------------
+% conditions are specified with `condition_requirement` and hold when
+% corresponding spec attributes hold.
+condition_holds(ID) :-
+  condition(ID);
+  attr(Name, A1)         : condition_requirement(ID, Name, A1);
+  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
+  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).
+
+% conditions that hold impose constraints on other specs
+attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
+attr(Name, A1, A2)     :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2).
+attr(Name, A1, A2, A3) :- condition_holds(ID), imposed_constraint(ID, Name, A1, A2, A3).
```

Comment:
The tests pass for me locally with the following diff:
```diff
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 1dfdf40cc3..3a643f9215 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -50,9 +50,22 @@ version_satisfies(Package, Constraint)
 % corresponding spec attributes hold.
 condition_holds(ID) :-
   condition(ID);
-  attr(Name, A1)         : condition_requirement(ID, Name, A1);
-  attr(Name, A1, A2)     : condition_requirement(ID, Name, A1, A2);
-  attr(Name, A1, A2, A3) : condition_requirement(ID, Name, A1, A2, A3).
+  attr(Name, A1)              : condition_requirement(ID, Name, A1);
+  attr(Name, A1, A2)          : condition_requirement(ID, Name, A1, A2);
+  attr(Name, A1, A2, A3)      : condition_requirement(ID, Name, A1, A2, A3);
+  % If this condition is a dependency, we need to have at least
+  % a dependency type defined
+  dependency_type_defined(ID) : dependency_condition(ID, _, _);
+  % Since a condition that holds on a dependency triggers the presence
+  % of the dependency, make it such that if the dependent is an external,
+  % then the condition doesn't hold
+  #false                      : dependency_of_an_external(ID).
+
+% Ensure we have at least a dependency type for dependency ID
+dependency_type_defined(ID) :- dependency_type(ID, _).
+
+% True if ID is a dependency condition and the dependent is an external
+dependency_of_an_external(ID) :- dependency_condition(ID, Package, _), external(Package).
 
 % conditions that hold impose constraints on other specs
 attr(Name, A1)         :- condition_holds(ID), imposed_constraint(ID, Name, A1).
```

Reply:
Taking a look at this -- I want to see if I can find a way to consolidate the logic here a bit more.  This sort of pollutes the general conditions with information about externals, but maybe there is no way around it...